## Problem Set II: Identify Areas Most Affected by The Top Complaint Type
### Reserach Question: Should the Department of Housing Preservation and Development of New York City focus on any particular set of boroughs, ZIP codes, or street (where the complaints are severe) for the specific type of complaints you identified in response to Question 1?

By: Dillon Morley

In [5]:
#Import Our Libraries
import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.

if os.environ.get('RUNTIME_ENV_LOCATION_TYPE') == 'external':
    endpoint_256cc9b3022041aeadc6f49afe94561b = 'https://s3.eu-geo.objectstorage.softlayer.net'
else:
    endpoint_256cc9b3022041aeadc6f49afe94561b = 'https://s3.eu-geo.objectstorage.service.networklayer.com'

client_256cc9b3022041aeadc6f49afe94561b = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='YvikRX53gzO8fvVreirDv5kHBltvbjmlUf5OWEVD7aSn',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url=endpoint_256cc9b3022041aeadc6f49afe94561b)

body = client_256cc9b3022041aeadc6f49afe94561b.get_object(Bucket='ibmpythondatasciencecapstone-donotdelete-pr-8kcbvnietwxfim',Key='311_Service_Requests_from_2010_to_Present_min.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

service_requests = pd.read_csv(body)
service_requests.head()


Unnamed: 0  Unique Key            Created Date Closed Date  \
0           0    45531130  02/02/2020 06:09:17 AM         NaN   
1           1    45529784  02/02/2020 02:15:24 PM         NaN   
2           2    45527528  02/02/2020 02:27:41 AM         NaN   
3           3    45530329  02/02/2020 12:13:18 PM         NaN   
4           4    45528814  02/02/2020 01:59:44 PM         NaN   

         Complaint Type         Location Type  Incident Zip  \
0        HEAT/HOT WATER  RESIDENTIAL BUILDING       10019.0   
1  UNSANITARY CONDITION  RESIDENTIAL BUILDING       11204.0   
2        HEAT/HOT WATER  RESIDENTIAL BUILDING       11372.0   
3        HEAT/HOT WATER  RESIDENTIAL BUILDING       10458.0   
4             APPLIANCE  RESIDENTIAL BUILDING       11209.0   

          Incident Address         Street Name Address Type             City  \
0     426 WEST   52 STREET    WEST   52 STREET      ADDRESS         NEW YORK   
1           1751 67 STREET           67 STREET      ADDRESS         BROOKLYN   
2          87-15 37 AVENUE           37 AVENUE      ADDRESS  Jackson Heights   
3  2405 SOUTHERN BOULEVARD  SOUTHERN BOULEVARD      ADDRESS            BRONX   
4            223 78 STREET           78 STREET      ADDRESS         BROOKLYN   

  Status                             Resolution Description    Borough  \
0   Open  The following complaint conditions are still o...  MANHATTAN   
1   Open  The following complaint conditions are still o...   BROOKLYN   
2   Open  The following complaint conditions are still o...     QUEENS   
3   Open  The following complaint conditions are still o...      BRONX   
4   Open  The following complaint conditions are still o...   BROOKLYN   

    Latitude  Longitude  
0  40.765132 -73.988993  
1  40.618484 -73.992673  
2  40.750269 -73.879432  
3  40.853773 -73.881558  
4  40.629745 -74.030533

In [6]:
#Function used to change the different column names to a standardized format.
def format_dataframe():
    service_requests.loc[service_requests['Complaint Type'] == 'PAINT - PLASTER'] = 'PAINT/PLASTER'
    service_requests.loc[service_requests['Complaint Type'] == 'Plumbing'] = 'PLUMBING'
    service_requests.loc[service_requests['Complaint Type'] == 'Electric'] = 'ELECTRIC'
    service_requests.loc[service_requests['Complaint Type'] == 'General'] = 'GENERAL'
    service_requests.loc[service_requests['Complaint Type'] == 'Safety'] = 'SAFETY'
    service_requests.loc[service_requests['Complaint Type'] == 'Appliance'] = 'APPLIANCE'
    service_requests.loc[service_requests['Complaint Type'] == 'Unsanitary Condition'] = 'UNSANITARY CONDITION'
    service_requests.loc[service_requests['Complaint Type'] == 'Outside Building'] = 'OUTSIDE BUILDING'
    service_requests.loc[service_requests['Complaint Type'] == 'HPD Literature Request'] = 'HPD LITERATURE REQUEST'
    service_requests.loc[service_requests['Complaint Type'] == 'Mold'] = 'MOLD'
    
format_dataframe()

print('These are are updated values for our complaint type column:',service_requests['Complaint Type'].unique())

These are are updated values for our complaint type column: ['HEAT/HOT WATER' 'UNSANITARY CONDITION' 'APPLIANCE' 'GENERAL' 'ELECTRIC'
 'PLUMBING' 'PAINT/PLASTER' 'WATER LEAK' 'SAFETY' 'DOOR/WINDOW'
 'OUTSIDE BUILDING' 'ELEVATOR' 'FLOORING/STAIRS' 'HPD LITERATURE REQUEST'
 'HEATING' 'GENERAL CONSTRUCTION' 'NONCONST' 'CONSTRUCTION' 'AGENCY'
 'STRUCTURAL' 'VACANT APARTMENT' 'MOLD']


### Question 1: For the complaint types that you selected in the previous module that had at least 800,000 complaints logged, which borough had the highest number of complaints submitted?

In [13]:
hot_water = service_requests.loc[service_requests['Complaint Type']=='HEAT/HOT WATER'] #Creating a dataframe for HEAT/HOT WATER complaints
print(hot_water.head(5))
heating = service_requests.loc[service_requests['Complaint Type']=='HEATING'] #Creating a dataframe for HEATING complaints
print(heating.head(5))

  Unnamed: 0 Unique Key            Created Date Closed Date  Complaint Type  \
0          0   45531130  02/02/2020 06:09:17 AM         NaN  HEAT/HOT WATER   
2          2   45527528  02/02/2020 02:27:41 AM         NaN  HEAT/HOT WATER   
3          3   45530329  02/02/2020 12:13:18 PM         NaN  HEAT/HOT WATER   
5          5   45530344  02/02/2020 07:52:08 AM         NaN  HEAT/HOT WATER   
7          7   45527939  02/02/2020 09:00:54 AM         NaN  HEAT/HOT WATER   

          Location Type Incident Zip         Incident Address  \
0  RESIDENTIAL BUILDING        10019     426 WEST   52 STREET   
2  RESIDENTIAL BUILDING        11372          87-15 37 AVENUE   
3  RESIDENTIAL BUILDING        10458  2405 SOUTHERN BOULEVARD   
5  RESIDENTIAL BUILDING        10456       1131 MORRIS AVENUE   
7  RESIDENTIAL BUILDING        11372          37-20 81 STREET   

          Street Name Address Type             City Status  \
0    WEST   52 STREET      ADDRESS         NEW YORK   Open   
2         

### Next step is to merge these two dataframes so we can see the combined dataframe for our top two complaints.

In [15]:
#Creates our merged dataframe and prints out the first 5 rows.
top_complaints = pd.merge(hot_water,heating,how='outer')
print(top_complaints.head(5))

   Unnamed: 0  Unique Key            Created Date Closed Date  Complaint Type  \
0         0.0  45531130.0  02/02/2020 06:09:17 AM         NaN  HEAT/HOT WATER   
1         2.0  45527528.0  02/02/2020 02:27:41 AM         NaN  HEAT/HOT WATER   
2         3.0  45530329.0  02/02/2020 12:13:18 PM         NaN  HEAT/HOT WATER   
3         5.0  45530344.0  02/02/2020 07:52:08 AM         NaN  HEAT/HOT WATER   
4         7.0  45527939.0  02/02/2020 09:00:54 AM         NaN  HEAT/HOT WATER   

          Location Type  Incident Zip         Incident Address  \
0  RESIDENTIAL BUILDING       10019.0     426 WEST   52 STREET   
1  RESIDENTIAL BUILDING       11372.0          87-15 37 AVENUE   
2  RESIDENTIAL BUILDING       10458.0  2405 SOUTHERN BOULEVARD   
3  RESIDENTIAL BUILDING       10456.0       1131 MORRIS AVENUE   
4  RESIDENTIAL BUILDING       11372.0          37-20 81 STREET   

          Street Name Address Type             City Status  \
0    WEST   52 STREET      ADDRESS         NEW YORK   

### Now we can use our value_counts( ) function to find the number of complaints from each borough in New York City.
As you can see the Bronx had the most complaints about Heating or Hot Water (606,123 complaints).

In [25]:
borough_complaints = top_complaints['Borough'].value_counts()
print('The Bronx had the most complaints about Heating or Hot Water with',borough_complaints[0],'complaints.')
print(borough_complaints)

The Bronx had the most complaints about Heating or Hot Water with 606133 complaints.
BRONX            606133
BROOKLYN         574845
MANHATTAN        423039
Unspecified      282917
QUEENS           244122
STATEN ISLAND     18368
Name: Borough, dtype: int64


### Question 2: For the complaint types that you selected in the previous module that had a total number that exceeded 800,000 complaints, which borough had the lowest number of complaints submitted?
Using the code above, we can easily see that Staten Island had the least amount of complaints about Heating or Hot Water (18,368 complaints).

In [27]:
print('Staten Island had the least amount of complaints about Heating or Hot Water with',borough_complaints[5],'complaints.')
print(borough_complaints)

Staten Island had the least amount of complaints about Heating or Hot Water with 18368 complaints.
BRONX            606133
BROOKLYN         574845
MANHATTAN        423039
Unspecified      282917
QUEENS           244122
STATEN ISLAND     18368
Name: Borough, dtype: int64


### Question 3: For the complaint types that you selected in the previous module that had at least 800,000 complaints logged, which ZIP code had the highest number of complainted submitted?
Using our merged dataframe and our value_counts( ) function we can easily find which zip code had the highest number of complaints submitted. First we need to convert our Incident Zip column to an integer type. However, there are NA values in our Zip Code column so we need to change those before converting our zip codes to an integer type.

In [43]:
top_complaints['Incident Zip'] = top_complaints['Incident Zip'].fillna(0) #Changes all the NA values to 0
top_complaints['Incident Zip'] = top_complaints['Incident Zip'].astype(int) #Changes all of our zip codes from float to integer type
zip_code_complaints = top_complaints['Incident Zip'].value_counts()
print('The zip code with the most complaints was 11226 with', zip_code_complaints[0],'complaints.')
print(zip_code_complaints)

The zip code with the most complaints was 11226 with 19024 complaints.
11226    69041
10467    66073
10458    65372
10468    58190
10453    57818
         ...  
10112        1
10121        1
10129        1
10045        1
10103        1
Name: Incident Zip, Length: 194, dtype: int64


### Question 4: T or F: For the complaint types that you selected in the previous module that had at least 800,000 complaints logged, the address 89-21 Elmhurst Avenue had the highest number of complainted submitted?

Using our merged dataframe and value_counts( ) we can easily find the street name that had the highest number of complaints submitted in our dataset.

In [48]:
address_complaints = top_complaints['Incident Address'].value_counts()
print('The address that filed the most complaints was 34 Arden Street with',address_complaints[0],'complaints filed.')
print(address_complaints)

The address that filed the most complaints was 34 Arden Street with 13858 complaints filed.
34 ARDEN STREET           13858
89-21 ELMHURST AVENUE     11548
3810 BAILEY AVENUE         7115
1025 BOYNTON AVENUE        5785
9511 SHORE ROAD            5006
                          ...  
2001 EAST 24 STREET           1
636 HOWARD AVENUE             1
110 BRIGHTON 11 STREET        1
1641 CASTLETON AVENUE         1
159-09 84 AVENUE              1
Name: Incident Address, Length: 136949, dtype: int64


The answer to question 4 is false. As you can see, the address: 34 Arden Street has the most complaints submitted in our dataset.

### Question 5: For the complaint types that you selected in the previous module that had at least 800,000 complaints logged, how many of the submitted tickets were closed?

Using our merged dataframe and value_counts() function, we can easily find the rows where submitted tickets = closed. 

In [55]:
results = top_complaints['Status'].value_counts()
print('There were',results[0],'closed complaints.')
print(results)

There were 2133331 closed complaints.
Closed     2133331
Open         16091
Pending          2
Name: Status, dtype: int64
